control-plane/pkg/reconciler/source/source.go

```diff
@@ -21,285 +21,163 @@ import (
-	// reconciliation loop.
-	InitOffsetsFunc kafka.InitOffsetsFunc
+type Reconciler struct {
+	ConsumerGroupLister internalslst.ConsumerGroupLister
+	InternalsClient     internalsclient.Interface
 }
 
 func (r *Reconciler) ReconcileKind(ctx context.Context, ks *sources.KafkaSource) reconciler.Event {
-	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
-		return r.reconcileKind(ctx, ks)
-	})
-}
 
-func (r *Reconciler) reconcileKind(ctx context.Context, ks *sources.KafkaSource) reconciler.Event {
-
-	logger := kafkalogging.CreateReconcileMethodLogger(ctx, ks)
-
-	statusConditionManager := base.StatusConditionManager{
-		Object:   ks,
-		Env:      r.Env,
-		Recorder: controller.GetEventRecorder(ctx),
-	}
-
-	if !r.IsDispatcherRunning() {
-		return statusConditionManager.DataPlaneNotAvailable()
-	}
-	statusConditionManager.DataPlaneAvailable()
-
-	authContext, err := security.ResolveAuthContextFromNetSpec(r.SecretLister, ks.GetNamespace(), ks.Spec.Net)
-	if err != nil {
-		return fmt.Errorf("failed to create auth context: %w", err)
-	}
-	secret, err := security.Secret(ctx, &SecretLocator{KafkaSource: ks}, security.NetSpecSecretProviderFunc(authContext))
+	cg, err := r.reconcileConsumerGroup(ctx, ks)
 	if err != nil {
-		return fmt.Errorf("failed to get secret: %w", err)
+		ks.GetConditionSet().Manage(&ks.Status).MarkFalse(KafkaConditionConsumerGroup, "failed to reconcile consumer group", err.Error())
+		return err
 	}
 
-	// get security option for Sarama with secret info in it
-	securityOption := security.NewSaramaSecurityOptionFromSecret(secret)
-
-	if err := security.TrackNetSpecSecrets(r.SecretTracker, ks.Spec.Net, ks); err != nil {
-		return fmt.Errorf("failed to track secrets: %w", err)
-	}
+	propagateConsumerGroupStatus(cg, ks)
 
-	saramaConfig, err := kafka.GetSaramaConfig(securityOption)
-	if err != nil {
-		return fmt.Errorf("error getting cluster admin sarama config: %w", err)
-	}
+	return nil
+}
 
-	kafkaClient, err := r.NewKafkaClient(ks.Spec.BootstrapServers, saramaConfig)
-	if err != nil {
-		return statusConditionManager.TopicsNotPresentOrInvalidErr(ks.Spec.Topics, fmt.Errorf("error getting sarama config: %w", err))
-	}
-	defer kafkaClient.Close()
+func (r Reconciler) reconcileConsumerGroup(ctx context.Context, ks *sources.KafkaSource) (*internalscg.ConsumerGroup, error) {
 
-	kafkaClusterAdminClient, err := r.NewKafkaClusterAdminClient(ks.Spec.BootstrapServers, saramaConfig)
-	if err != nil {
-		return statusConditionManager.TopicsNotPresentOrInvalidErr(ks.Spec.Topics, fmt.Errorf("cannot obtain Kafka cluster admin, %w", err))
-	}
-	defer kafkaClusterAdminClient.Close()
+	backoffPolicy := eventingduck.BackoffPolicyExponential
 
-	isValid, err := kafka.AreTopicsPresentAndValid(kafkaClusterAdminClient, ks.Spec.Topics...)
-	if err != nil {
-		return statusConditionManager.TopicsNotPresentOrInvalidErr(ks.Spec.Topics, err)
-	}
-	if !isValid {
-		return statusConditionManager.TopicsNotPresentOrInvalid(ks.Spec.Topics)
```

Comment:
Hey @pierDipi for kafkaSource, shouldn't we have this code to check topic validity and existence for the new scaling source as well?

Reply:
That's a good check to add to the ConsumerGroup reconciler since there we create the admin client